extensions/amp-a4a/0.1/refresh-manager.js

```diff
@@ -0,0 +1,364 @@
+ * same IO.
+ *
+ * @private {!Object<string, (!IntersectionObserver|!IntersectionObserverPolyfill)>}
+ */
+const observers = {};
+
+export class RefreshManager {
+
+  /**
+   * @param {!./amp-a4a.AmpA4A} a4a The AmpA4A instance to be refreshed.
+   * @param {!RefreshConfig} config
+   */
+  constructor(a4a, config) {
+
+    /** @private {string} */
+    this.state_ = RefreshLifecycleState.INITIAL;
+
+    /** @const @private {!./amp-a4a.AmpA4A} */
+    this.a4a_ = a4a;
+
+    /** @const @private {!Window} */
+    this.win_ = a4a.win;
+
+    /** @const @private {!Element} */
+    this.element_ = a4a.element;
+
+    /** @const @private {string} */
+    this.adType_ = this.element_.getAttribute('type').toLowerCase();
+
+    /** @const @private {?number} */
+    this.refreshInterval_ = this.getPublisherSpecifiedRefreshInterval_();
+
+    /** @const @private {!RefreshConfig} */
+    this.config_ = this.convertConfiguration_(config);
+
+    /** @const @private {!../../../src/service/timer-impl.Timer} */
+    this.timer_ = timerFor(this.win_);
+
+    /** @private {?(number|string)} */
+    this.refreshTimeoutId_ = null;
+
+    /** @private {?(number|string)} */
+    this.visibilityTimeoutId_ = null;
+
+    /** @private {boolean} */
+    this.isRefreshable_ = isExperimentOn(this.win_, 'amp-ad-refresh') &&
+        // The network has opted-in.
+        !!(this.config_
+        // The publisher has enabled refresh on this slot.
+        && (this.refreshInterval_ || this.refreshInterval_ == '')
+        // The slot is contained only within container types eligible for
+        // refresh.
+        && !getEnclosingContainerTypes(this.element_).filter(container =>
+          container != ValidAdContainerTypes['AMP-CAROUSEL']
+          && container != ValidAdContainerTypes['AMP-STICKY-AD']).length);
+
+    if (this.isRefreshable_) {
+      this.initiateRefreshCycle();
+    }
+  }
+
+  /**
+   * Returns an IntersectionObserver configured to the given threshold, creating
+   * one if one does not yet exist.
+   *
+   * @param {(string|number)} threshold
+   * @return {(!IntersectionObserver|!IntersectionObserverPolyfill)}
+   */
+  getIntersectionObserverWithThreshold_(threshold) {
+    threshold = String(threshold);
+    return observers[threshold] ||
+        (observers[threshold] = 'IntersectionObserver' in this.win_
+         ? new this.win_['IntersectionObserver'](
+             this.ioCallbackGenerator_(), {threshold})
+         : new IntersectionObserverPolyfill(
+             this.ioCallbackGenerator_(), {threshold}));
+  }
+
+  /**
+   * Returns a function that will be invoked directly by the
+   * IntersectionObserver implementation. It will implement the core logic of
+   * the refresh lifecycle, including the transitions of the DFA.
+   *
+   * @return {function(!Array<IntersectionObserverEntry>)}
+   */
+  ioCallbackGenerator_() {
+    const refreshManager = this;
+    return entries => {
```

Comment:
Any reason this can't be:
return entries => entries.forEach(entry => {
  if (entry.target != refreshManager.element_) return;
  switch(refreshManager.state_) {
    ...
  }
});